# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/RegExpObject.cpp
# Commit: 410d1b57bb09
# Full Hash: 410d1b57bb0955e0fe2f47b95ffa5ee1005e6791
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-09 21:37:17
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 5: Replace heap state checks that rely on there being a JSContext for the thread r=jandem
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D140192
# ==============================================================================

diff -r f78d139e1d0e -r 410d1b57bb09 js/src/vm/RegExpObject.cpp
--- a/js/src/vm/RegExpObject.cpp	Wed Mar 09 10:17:07 2022 +0000
+++ b/js/src/vm/RegExpObject.cpp	Wed Mar 09 10:17:07 2022 +0000
@@ -153,18 +153,6 @@
   return false;
 }
 
-static inline bool IsMarkingTrace(JSTracer* trc) {
-  // Determine whether tracing is happening during normal marking.  We need to
-  // test all the following conditions, since:
-  //
-  //   1. During TraceRuntime, RuntimeHeapIsBusy() is true, but the
-  //      tracer might not be a marking tracer.
-  //   2. When a write barrier executes, isMarkingTracer is true, but
-  //      RuntimeHeapIsBusy() will be false.
-
-  return JS::RuntimeHeapIsCollecting() && trc->isMarkingTracer();
-}
-
 static const ClassSpec RegExpObjectClassSpec = {
     GenericCreateConstructor<js::regexp_construct, 2, gc::AllocKind::FUNCTION>,
     GenericCreatePrototype<RegExpObject>,
@@ -556,7 +544,11 @@
 
 void RegExpShared::traceChildren(JSTracer* trc) {
   // Discard code to avoid holding onto ExecutablePools.
-  if (IsMarkingTrace(trc) && trc->runtime()->gc.isShrinkingGC()) {
+  gc::GCRuntime* gc = &trc->runtime()->gc;
+  // Bug 1758095: isIncrementalGCInProgress() also returns true when called from
+  // the collector during a non-incremental GC.
+  if (trc->isMarkingTracer() && gc->isIncrementalGCInProgress() &&
+      gc->isShrinkingGC()) {
     discardJitCode();
   }
 
